### PR TITLE
GEM Intial Feedback fixes

### LINKED ISSUE #19156: ISIS_Powder - GEM Initial feedback fixes
Following an initial demonstration the following issues were noticed

- [x] `Friendly_enum_name` is missing for GEM modes (check all)
- [x] The YAML is case sensitive from the mode set on the script (e.g. `Rietveld` requires the same caps in the YAML file)
- [x] `run_in_range` ends up being used for the output name (and saved file name). I've seen this on POLARIS too
- [x] When `unit_to_keep` is set to `d-spacing` it incorrectly rejects it with the same output type name
- [x] Number density for V Nb rod was set incorrectly 
- [x] Input batching option for Gem
- [ ] When an enum is missing (checked with enum_class) we should print acceptable values too

### PR DESCRIPTION
**Pull request by DavidFair**
~~**Note cannot be merged before #19163**~~

**Description of work.**
This PR adds the GEM instrument fixes to ISIS Powder 
Additionally it adds a generic method where we can pass functions to specialise the properties held in a run details object on a case by case basis 

**To test:**
Code review
Ensure unit tests pass

Fixes #19156 .

*Does not need to be in the release notes as currently in development*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [x] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/test/ISISPowderRunDetailsTest.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+    def test_run_details_splined_name_list_is_used(self):
+        expected_vanadium_runs = "11-12"
+        expected_offset_file_name = "offset_file_name"
+        splined_name_list = ["bar", "bang", "baz"]
+        mock_inst = self.setup_mock_inst_settings(yaml_file_path="ISISPowderRunDetailsTest.yaml")
+        output_obj = RunDetails.create_run_details_object(run_number_string=10, inst_settings=mock_inst,
+                                                          is_vanadium_run=False, splined_name_list=splined_name_list)
+
+        expected_splined_out_str = ''.join('_' + val for val in splined_name_list)
```
Comment (DimitarTasev) [inline scripts/test/ISISPowderRunDetailsTest.py]: Adding `[ ]` to make it a list -> **[**"_" + val for val in splined_name_list **]** is about 2x faster
```
In [62]: timeit('"".join(["_" + val for val in splined_name_list])',setup='splined_name_list = ["name","2","fff"]')
Out[62]: 0.4237639904022217
```
```
In [63]: timeit('"".join("_" + val for val in splined_name_list)',setup='splined_name_list = ["name","2","fff"]')
Out[63]: 0.9818799495697021
```

File: scripts/test/ISISPowderRunDetailsTest.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+class MockInstSettings(object):
+    def __init__(self, cal_file_path, calibration_dir):
+        self.calibration_dir = calibration_dir
+        self.cal_mapping_path = cal_file_path
+        self.grouping_file_name = MockInstSettings.gen_random_string()
+
+    @staticmethod
+    def gen_random_string():
+        return ''.join(random.choice(string.ascii_lowercase) for _ in range(10))
```
Comment (DimitarTasev) [inline scripts/test/ISISPowderRunDetailsTest.py]: Can save another 0.5s per call here by making the code inside `join(...)` a list:

```
In [66]: timeit("''.join(random.choice(string.ascii_lowercase) for _ in range(10))",setup='import random, string; splined_name_list = ["name","2","fff"]')
Out[66]: 6.115681886672974
```
```
In [67]: timeit("''.join([random.choice(string.ascii_lowercase) for _ in range(10)])",setup='import random, string; splined_name_list = ["name","2","fff"]')
Out[67]: 5.514591932296753
```

**Review by dtasev — CHANGES_REQUESTED** [review]

**Review by dtasev — APPROVED** [review]

**Comment by DimitarTasev** [comment]
Everything looks good, unit tests pass, when the build servers pass :shipit:

**Comment by DTasev** [comment]
changes look good :shipit:

**Pull request merged**